src/pass/tensor_core.cc

```diff
@@ -0,0 +1,1247 @@
+                      double cuda_compute_capability, double cuda_version) {
+    for (auto kv : extern_buffer) {
+      BufferInfo bi;
+      bi.name = kv.second->name;
+      bi.dtype = kv.second->dtype;
+      bi.external = true;
+      buf_map_[TensorKey{kv.first->op, kv.first->value_index}] = bi;
+    }
+    // Int wmma is supported when cuda version >= 10.0 && cuda arch >= 720
+    if (cuda_compute_capability >= 7.20 && cuda_version >= 10.0) {
+      support_int_wmma_ = true;
+    }
+  }
+  using IRVisitor::Visit_;
+
+  void Visit_(const AttrStmt* op) final {
+    if (op->attr_key == attr::pragma_tensor_core) {
+      tensor_core_on_ = true;
+      IRVisitor::Visit_(op);
+    } else if (op->attr_key == attr::realize_scope) {
+      storage_scope_[op->node.get()] = op->value.as<StringImm>()->value;
+      Visit(op->body);
+    } else {
+      IRVisitor::Visit_(op);
+    }
+  }
+
+  void Visit_(const Provide* op) final {
+    IRVisitor::Visit_(op);
+    auto it = buf_map_.find(TensorKey{op->func, op->value_index});
+    if (it == buf_map_.end()) {
+      return;
+    }
+    const BufferInfo& bi = it->second;
+    if (bi.released) {
+      return;
+    }
+    if (tensor_core_on_ && mma_sync_match_(op, bi)) {
+      matched_ = true;
+    }
+  }
+
+  void Visit_(const Realize* op) final {
+    TensorKey key{op->func, op->value_index};
+    if (buf_map_.count(key)) {
+      if (!buf_map_.at(key).external) {
+        return;
+      }
+      Visit(op->body);
+    } else {
+      BufferInfo bi;
+      bi.name = key.GetName();
+      bi.dtype = op->type;
+      buf_map_[key] = bi;
+      Visit(op->body);
+      buf_map_[key].released = true;
+    }
+  }
+
+  inline bool Matched() const {return matched_;}
+
+  friend class ScheduleAnalyser;
+  friend class BufferAnalyser;
+
+ private:
+  struct BufferInfo {
+    std::string name;
+    Type dtype;
+    bool external{false};
+    bool released{false};
+    bool same_as(const BufferInfo &bi) {
+      if (this->dtype != bi.dtype) return false;
+      if (this->name != bi.name) return false;
+      if (this->external != bi.external) return false;
+      if (this->released != bi.released) return false;
+      return true;
+    }
+  };
+
+  // Check whether the storage scope is local
+  bool check_local_buffer_(const Call* op, BufferInfo* bi) {
+    if (op->call_type == Call::Halide) {
+      auto it = storage_scope_.find(op->func.get());
+      if (it == storage_scope_.end()) {
+        return false;
+      }
+      const std::string& strkey = it->second;
+      if (strkey != "local") {
+        return false;
+      }
+      auto it1 = buf_map_.find(TensorKey{op->func, op->value_index});
+      if (it1 == buf_map_.end()) {
+        return false;
+      }
+      *bi = it1->second;
+      if (bi->released) {
+        return false;
+      }
+      return true;
+    }
+    return false;
+  }
+
+  // Do the pattern matching
+  bool mma_sync_match_(const Provide* op, BufferInfo store_buffer) {
```

Comment:
Dumb question, does this function looks for wmma and replace with mma.sync?

Reply:
This part is still in analysing phase. We do the mma pattern match here as well as record some matrix info.
And yes, finally we'll replace the whole AST block with a mma.sync Intrinsic call.

Reply:
Awesome. If it is not too much trouble can you dump ptx into a gist and paste the link here? I can review it and suggest few changes if needed

Reply:
> Awesome. If it is not too much trouble can you dump ptx into a gist and paste the link here? I can review it and suggest few changes if needed
Thanks for the kindness. BTW, what do you expect to get by looking at the ptx assembly?
For cuda, one single mma.sync C APi will be replaced with several PTX instructions via nvcc, so may be you are afraid that there may be some miss-usage?

Reply:
That is one reason. Here is why I want the ptx dump
1. I want to see if the schedule is good or not. If not, I can suggest how it should be. May be we can change few things to get it right
2. To see if there are any low-throughput ptx instructions that are causing any slowdown.
3. Want to see how shared memory is being used. mma.sync requires input operands (a, b and c) to be laid out in specific pattern. This can cause shared memory bank conflicts.

For your question. What do you mean by C API? Is it the C wrapper around intrinsic?

Reply:
We generate CUDA code instead of PTX code, so it's nvcc that decides which ptx instructions to use. Yes, we are continuously optimizing the schedule, and any better schedule is also welcome. But it is beyond the scope of this pull request. The main goal of this pull request is not to deliver the best schedules, but a feature and a tutorial to guide how to use this feature and to prove it can achieve good enough performance in some cases (at least be able to reproduce the results in the RFC). If you are interested, we are looking forward to corporating with you to deliver better and better schedules. Thanks!

Reply:
> Awesome. If it is not too much trouble can you dump ptx into a gist and paste the link here? I can review it and suggest few changes if needed

Hi Aditya, I have sent you the generated cuda code as well as the ptx compiled with nvcc via a Message at https://discuss.tvm.ai/. Could you please help to take a look? Thank you!